### PR TITLE
Add config entry to choose allowed image dimensions.

### PR DESCRIPTION
**Pull request by debonzi**
When ALLOW_DIMENSIONS is set as list of image dimensions (e.g ['100x100', '200x200']=] it will return 404 for all dimensions but the ones defined.

### CODE REVIEW

**Comment by guilhermef** [comment]
@debonzi Why do you need to limit the images size ?
If you generate a signed url, nobody who doesn't have the key will be able to generate another size.

**Comment by debonzi** [comment]
@guilhermef 
I intend to use it on mobile clients and, as far as I can see, I would need to ship the app with the KEY embedded which could make some trouble if someone manage to retrieve it and/or I need for any reason to change it on server side. Also, I will have a full JS web interface that would expose the KEY unless I create a proxy server to hold it and implement the dimensions filter at it.
I considered to use the blacklist that is already implemented but the functionality I need is more like a whitelist.

**Comment by guilhermef** [comment]
@dhardy92 @heynemann What do you guys think ?
It's looking like a valid scenario.

@debonzi do you mind adding some test cases for this functionality ?

**Comment by debonzi** [comment]
@guilhermef 
No problem. Working on it.

**Comment by guilhermef** [comment]
@scorphus we need your opinion too.

**Comment by heynemann** [comment]
I am for accepting this PR. Looks like a valid use. Couple things I'd add before accepting this PR:

* Change the default thumbor.conf to include this configuration commented out with the default value and the comment explaining how to use it;
* We need to define who's going to update the wiki with this new info. Do you think you can do it @debonzi? There are a couple pages to update.

**Comment by scorphus** [comment]
I imagine @debonzi's intent with this PR is to prevent URL tampering. But how about other available options such as filters? One could still overload the service with `fill()`, for instance:

    http://example.com/unsafe/200x200/filters:fill(ffffff,true)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(ffffff,false)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(ffffff,1)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(ffffff,0)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(fffffe,true)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(fffffe,false)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(fffffe,1)/path/to/picture.png
    http://example.com/unsafe/200x200/filters:fill(fffffe,0)/path/to/picture.png

File: thumbor/config.py [hunk 1]
```diff
@@ -74,6 +74,10 @@
     'Indicates whether thumbor should enable blacklist functionality to prevent processing certain images.', 'Imaging')
 
 Config.define(
+    'ALLOW_DIMENSIONS', [],
+    'Indicates whether thumbor should process certain images dimensions.', 'Imaging')
```
Comment (crccheck) [inline thumbor/config.py]: To me, it seems like this should read "certain image dimensions"

**Comment by dhardy92** [comment]
I agree with @scorphus this is not a real protection (just like spreading secret key, better using /unsafe/ actually)
Would be better filter at load balancing level (apache, haproxy,nginx, F5...) with rules allowing Location containg only patterns matching what is really wanted and 403 for other.

**Comment by guilhermef** [comment]
I agree with them, that wouldn't avoid url tampering.
@debonzi how about @dhardy92 solution ?

**Comment by debonzi** [comment]
Hi all,
@guilhermef  I didn't realize that it would still be possible to tamper using fill() because I am not using filtering.
First plan was to use @dhardy92 solution, but it seamed to me that it would be a nice option to be available on Thumbor. Another motivation is that I will use AWS Deis (Thumbor is running in a Docker container) and nginx, LB, etc are already on Deis environment working out of the box. I am not sure I can do those custom configurations easily on those services but been able to configure it on Thumbor, a simple 'git push' on deis infrastructure would do the magic.

I don't see an easy way to solve the fill() (and possible other) way to tamper with Thumbor url so I guess I will need to find a way to use on of the proposed solutions.

Thank you all anyway.

**Comment by RTodorov** [comment]
+1 on this. Whitelist is a very useful feature, although I can't see a clear solution to the problem with the filters. To me, if someone wants to use the whitelist for security, he should also block or limit the use of filters.

**Comment by debonzi** [comment]
I did another commit where it will return 403 when using filters with ALLOW_DIMENSIONS set. Also, described it on configuration.
If filters are the only case where it would be a flaw, I think it would be covered now.
Do you guys think it is acceptable?

**Comment by guilhermef** [comment]
@debonzi I think It would be better to have a separated config like, DISABLE_FILTERS.

It's kinda strange to have ALLOW_DIMENSIONS disabling filters.

**Comment by debonzi** [comment]
@guilhermef that is a way to go too, but my intention with that was to avoid someone to set ALLOW_DIMENSIONS and believe it safe when it is not.
If you guys think that doing a DISABLE_FILTERS and just suggest its use with ALLOW_DIMENSIONS on config comment and future documentation for this is ok, that would work for me too.

**Comment by masom** [comment]
@debonzi Why not just disable filters with an empty list in FILTERS? Same goes with Optimizers or any future features that could potentially be "unsafe".

`ALLOWED_DIMENSIONS` is quite handy if you want to limit what image sizes web developers are allowed to use vs allowing all sizes. I don't think it should prevent filters from being used.

**Comment by debonzi** [comment]
@masom That is fine for me too. All I need is to be able to limit the image sizes allowed. I just did this change based on @scorphus comment.
I am happy making any changes you guys think is best to get it merged.

**Comment by felipemorais** [comment]
+1 ALLOWED_DIMENSIONS
+1 ALLOWED_FILTERS

**Comment by masom** [comment]
@felipemorais `ALLOWED_FILTERS` is already implement as `FILTERS`

**Comment by jeremyk** [comment]
Another suggestion to solve this problem (that would be a larger change) is to have WHITELISTED_OPERATIONS where you could have a name like 'default_thumbnail' that defines the size/filters part of the url. The url instead of /unsafe/200x200/filters/path_to_image would just be /named/default_thumbnail/path_to_image. The thumbor would just replace /named/default_thumbnail/ with the transformation. Has some caching implications when changing operation on the server but seems secure.

**Comment by guilhermef** [comment]
The problem is that even with the FILTERS conf empty,
some attacker would still be able to request something like this:

http://example.com/unsafe/200x200/filters:fill(ffffff,true)/path/to/picture.png

The url regex still treats this url as valid.

**Comment by heynemann** [comment]
@guilhermef is right. It's not a problem of actually executing the filters, since thumbor is not the issue here. It will just ignore filters that do not exist. The actual problem is caching. Since the URLs generated can be infinite, an attacker will be able to:

a) Fill the cache with infinite objects (all the same image, though, since filters will be ignored);
b) Overload thumbor since the caching mechanism will be 100% misses.

Both are equally bad for any given site, or app for that matter.

This is a hard problem, guys. Do not be disheartened because we haven't found a solution for it. Most imaging solutions haven't.

One possible solution would be to have a third route, something like `/preset/small-thumb/source-image.jpg`. This is a route that does not require the private key. This is the way imgix and cloudinary support both Front-end and Apps images. The presets need to be, well, pre set, as in you have to have them in thumbor's configuration. 

The key thing to understand here is that these presets include all available options to thumbor. You won't be able to use a preset with an URL specified filter. You will be able to use presets in more aggressive environments though (browser, apps, etc).

What do you guys think?

**Comment by jeremyk** [comment]
@heynemann - I think this is what I was suggesting above. Right?

**Comment by heynemann** [comment]
Yep. Sorry for repeating it. We just need to ensure it uses a different route than the signed urls. I think we should close this PR and open a new one with this change. What do you think?

**Comment by jeremyk** [comment]
No worries, just wanted to make sure. Anyway, that sounds good to me!

**Comment by scorphus** [comment]
It's been almost a year since we last discussed about this pull request. Some nice ideas came out of the initial suggestion. Should anyone like to see them in Thumbor, just open a new PR and we'll kindly review it. I'm closing this one for now.

**Pull request closed without merging**